File: q2/q2java/Engine.c
#include "globals.h"  // to get at the game_import_t structure
#include "javalink.h"  // for some helper functions
#include "Engine.h"


#define CALL_MULTICAST 0
#define CALL_WRITEPOSITION 1
#define CALL_WRITEDIR 2
#define CALL_WRITECHAR 3
#define CALL_WRITEBYTE 4
#define CALL_WRITESHORT 5
#define CALL_WRITELONG 6
#define CALL_WRITEFLOAT 7
#define CALL_WRITEANGLE 8
#define CALL_WRITESTRING 9
#define CALL_UNICAST 10
#define CALL_INPVS 11
#define CALL_INPHS 12


// handle to Engine class
static jclass class_Engine;

// handle to TraceResults class
static jclass class_TraceResults;
static jmethodID method_TraceResults_ctor;

static JNINativeMethod Engine_methods[] = 
	{
	{"dprint",		"(Ljava/lang/String;)V", 		Java_Engine_dprint},
	{"bprint",		"(ILjava/lang/String;)V",		Java_Engine_bprint},
	{"configString","(ILjava/lang/String;)V",		Java_Engine_configString},
	{"error",		"(Ljava/lang/String;)V",		Java_Engine_error},
	{"modelIndex",	"(Ljava/lang/String;)I",		Java_Engine_modelIndex},
	{"soundIndex",	"(Ljava/lang/String;)I",		Java_Engine_soundIndex},
	{"imageIndex",	"(Ljava/lang/String;)I",		Java_Engine_imageIndex},
	{"trace0",		"(FFFFFFFFFFFFLq2java/NativeEntity;I)Lq2java/TraceResults;",	Java_Engine_trace0},
	{"pointContents0",		"(FFF)I",				Java_Engine_pointContents0},
	{"inP0",				"(FFFFFFI)Z",			Java_Engine_inP0},
	{"setAreaPortalState",	"(IZ)V",				Java_Engine_setAreaPortalState},
	{"areasConnected",		"(II)Z",				Java_Engine_areasConnected},
	{"boxEntities0",		"(FFFFFFII)[Lq2java/NativeEntity;",	Java_Engine_boxEntities0},
	{"write0",		"(Ljava/lang/Object;FFFII)V",	Java_Engine_write0},
	{"argc",		"()I", 							Java_Engine_argc},
	{"argv",		"(I)Ljava/lang/String;",		Java_Engine_argv},
	{"args",		"()Ljava/lang/String;",			Java_Engine_args},
	{"addCommandString",	"(Ljava/lang/String;)V",Java_Engine_addCommandString},
	{"debugGraph",			"(FI)V",				Java_Engine_debugGraph},
	{"getGamePath",			"()Ljava/lang/String;",	Java_Engine_getGamePath}
	};


void Engine_javaInit()
	{
	class_Engine = (*java_env)->FindClass(java_env, "q2java/Engine");
	CHECK_EXCEPTION();
	if (!class_Engine)
		{
		debugLog("Couldn't get Java Engine class\n");
		return;
		}

	(*java_env)->RegisterNatives(java_env, class_Engine, Engine_methods, 20);
	CHECK_EXCEPTION();

	
	class_TraceResults = (*java_env)->FindClass(java_env, "q2java/TraceResults");
	CHECK_EXCEPTION();
	if (!class_TraceResults)
		{
		debugLog("Couldn't get Java TraceResults class\n");
		return;
		}

	method_TraceResults_ctor = (*java_env)->GetMethodID(java_env, class_TraceResults, "<init>", "(ZZFLq2java/Vec3;Lq2java/Vec3;FBBLjava/lang/String;IIILq2java/NativeEntity;)V");
	CHECK_EXCEPTION();
	if (!method_TraceResults_ctor)
		{
		debugLog("Couldn't get Java TraceResults constructor\n");
		return;
		}
	}

void Engine_javaFinalize()
	{
	if (class_Engine)
		(*java_env)->UnregisterNatives(java_env, class_Engine); 
	}

static void JNICALL Java_Engine_dprint(JNIEnv *env, jclass cls, jstring s)
	{
	const char *str;

	str = (*env)->GetStringUTFChars(env, s, 0);
	gi.dprintf("%s", str); 
	debugLog(str);
	(*env)->ReleaseStringUTFChars(env, s, str);
	}

static void JNICALL Java_Engine_bprint(JNIEnv *env, jclass cls, jint printlevel, jstring s)
	{
	const char *str;

	str = (*env)->GetStringUTFChars(env, s, 0);
	gi.bprintf(printlevel, "%s", str); 
	(*env)->ReleaseStringUTFChars(env, s, str);
	}

static void JNICALL Java_Engine_configString(JNIEnv *env, jclass cls, jint num, jstring js)
	{
	char *s;

	s = (char *)((*env)->GetStringUTFChars(env, js, 0));
	gi.configstring(num, s);
	(*env)->ReleaseStringUTFChars(env, js, s);
	}


static void JNICALL Java_Engine_error(JNIEnv *env, jclass cls, jstring js)
	{
	char *s;

	s = (char *)((*env)->GetStringUTFChars(env, js, 0));
	gi.error(s);
	(*env)->ReleaseStringUTFChars(env, js, s);
	}


static jint JNICALL Java_Engine_modelIndex(JNIEnv *env, jclass cls, jstring jname)
	{
	char *name;
	int result;

	name = (char *)((*env)->GetStringUTFChars(env, jname, 0));
	result = gi.modelindex(name);
	(*env)->ReleaseStringUTFChars(env, jname, name);
	return result;
	}

static jint JNICALL Java_Engine_soundIndex(JNIEnv *env, jclass cls, jstring jname)
	{
	char *name;
	int result;

	name = (char *)((*env)->GetStringUTFChars(env, jname, 0));
	result = gi.soundindex(name);
	(*env)->ReleaseStringUTFChars(env, jname, name);
	return result;
	}

static jint JNICALL Java_Engine_imageIndex(JNIEnv *env, jclass cls, jstring jname)
	{
	char *name;
	int result;

	name = (char *)((*env)->GetStringUTFChars(env, jname, 0));
	result = gi.imageindex(name);
	(*env)->ReleaseStringUTFChars(env, jname, name);
	return result;
	}


static jobject JNICALL Java_Engine_trace0(JNIEnv *env, jclass cls, 
	jfloat startx, jfloat starty, jfloat startz, 
	jfloat minsx, jfloat minsy, jfloat minsz, 
	jfloat maxsx, jfloat maxsy, jfloat maxsz, 
	jfloat endx, jfloat endy, jfloat endz, 
	jobject jpassEnt, jint contentMask)
	{
	vec3_t start;
	vec3_t mins;
	vec3_t maxs;
	vec3_t end;
	edict_t *passEnt;

	trace_t result;

	jobject resEndpos;
	jobject resPlaneNormal;
	jstring resSurfaceName;
	int resSurfaceFlags;
	int resSurfaceValue;
	jobject resEnt;

	start[0] = startx;
	start[1] = starty;
	start[2] = startz;

	mins[0] = minsx;
	mins[1] = minsy;
	mins[2] = minsz;

	maxs[0] = maxsx;
	maxs[1] = maxsy;
	maxs[2] = maxsz;

	end[0] = endx;
	end[1] = endy;
	end[2] = endz;

	passEnt = ge.edicts + Entity_get_fEntityIndex(jpassEnt);

	result = gi.trace(start, mins, maxs, end, passEnt, contentMask);

	resEndpos = newJavaVec3(&(result.endpos));
	resPlaneNormal = newJavaVec3(&(result.plane.normal));

	if (!result.surface)
		{
		resSurfaceName = 0;
		resSurfaceFlags = resSurfaceValue = 0;
		}
	else
		{
		resSurfaceName = (*env)->NewStringUTF(env, result.surface->name);
		resSurfaceFlags = result.surface->flags;
		resSurfaceValue = result.surface->value;
		}

	resEnt = Entity_getEntity(result.ent - ge.edicts);

	return (*env)->NewObject(java_env, class_TraceResults, method_TraceResults_ctor, 
		result.allsolid, result.startsolid, result.fraction, resEndpos, resPlaneNormal, 
		result.plane.dist, result.plane.type, result.plane.signbits,
		resSurfaceName, resSurfaceFlags, resSurfaceValue, result.contents,
		resEnt);
	}


static jint JNICALL Java_Engine_pointContents0(JNIEnv *env, jclass cls, jfloat x, jfloat y, jfloat z)
	{
	vec3_t v;
	v[0] = x;
	v[1] = y;
	v[2] = z;
	return gi.pointcontents(v);
	}

static jboolean JNICALL Java_Engine_inP0(JNIEnv *env, jclass cls, jfloat x1, jfloat y1, jfloat z1, jfloat x2, jfloat y2, jfloat z2, jint calltype)
	{
	vec3_t p1, p2;
	p1[0] = x1;
	p1[1] = y1;
	p1[2] = z1;
	p2[0] = x2;
	p2[1] = y2;
	p2[2] = z2;
	switch (calltype)
		{
		case CALL_INPHS: return gi.inPHS(p1, p2);
		case CALL_INPVS: return gi.inPVS(p1, p2);
		default: return 0;
		}
	}


static void JNICALL Java_Engine_setAreaPortalState(JNIEnv *env, jclass cls, jint portalnum, jboolean open)
	{
	gi.SetAreaPortalState(portalnum, open);
	}


static jboolean JNICALL Java_Engine_areasConnected(JNIEnv *env, jclass cls, jint area1, jint area2)
	{
	return gi.AreasConnected(area1, area2);
	}


static jobjectArray JNICALL Java_Engine_boxEntities0(JNIEnv *env, jclass cls, 
	jfloat minsx, jfloat minsy, jfloat minsz, 
	jfloat maxsx, jfloat maxsy, jfloat maxsz, jint maxCount, jint areaType)
	{
	vec3_t mins;
	vec3_t maxs;
	int count;
	edict_t **list = gi.TagMalloc(maxCount * sizeof(edict_t *), TAG_GAME);
	jobjectArray result;

	mins[0] = minsx;
	mins[1] = minsy;
	mins[2] = minsz;

	maxs[0] = maxsx;
	maxs[1] = maxsy;
	maxs[2] = maxsz;

	count = gi.BoxEdicts(mins, maxs, list, maxCount, areaType);

	result = Entity_createArray(list, count);
	gi.TagFree(list);

	return result;
	}


static void JNICALL Java_Engine_write0(JNIEnv *env, jclass cls, jobject obj, jfloat x, jfloat y, jfloat z, jint c, jint calltype)
	{
	vec3_t v;
	char *s;
	int index;

	switch (calltype)
		{
		case CALL_MULTICAST:
			v[0] = x;
			v[1] = y;
			v[2] = z;
			gi.multicast(v, c);
			break;

		case CALL_UNICAST:
			index = Entity_get_fEntityIndex(obj);
			if ((index >= 0) && (index < ge.max_edicts))
				gi.unicast(ge.edicts + index, c);
			break;

		case CALL_WRITECHAR:
			gi.WriteChar(c);
			break;

		case CALL_WRITEBYTE:
			gi.WriteByte(c);
			break;

		case CALL_WRITESHORT:
			gi.WriteShort(c);
			break;

		case CALL_WRITELONG:
			gi.WriteLong(c);
			break;

		case CALL_WRITEFLOAT:
			gi.WriteFloat(x);
			break;

		case CALL_WRITEANGLE:
			gi.WriteAngle(x);
			break;

		case CALL_WRITESTRING:
			s = (char *)((*env)->GetStringUTFChars(env, obj, 0));
			gi.WriteString(s);
			(*env)->ReleaseStringUTFChars(env, obj, s);
			break;

		case CALL_WRITEPOSITION:
			v[0] = x;
			v[1] = y;
			v[2] = z;
			gi.WritePosition(v);
			break;

		case CALL_WRITEDIR:
			v[0] = x;
			v[1] = y;
			v[2] = z;
			gi.WriteDir(v);
			break;
		}
	}

static jint JNICALL Java_Engine_argc(JNIEnv *env, jclass cls)
	{
	return gi.argc();	
	}

static jstring JNICALL Java_Engine_argv(JNIEnv *env, jclass cls, jint n)
	{
	return (*env)->NewStringUTF(env, gi.argv(n));
	}

static jstring JNICALL Java_Engine_args(JNIEnv *env, jclass cls)
	{
	return (*env)->NewStringUTF(env, gi.args());
	}

static void JNICALL Java_Engine_addCommandString(JNIEnv *env, jclass cls, jstring js)
	{
	char *s;

	s = (char *)((*env)->GetStringUTFChars(env, js, 0));
	gi.AddCommandString(s);
	(*env)->ReleaseStringUTFChars(env, js, s);
	}

static void JNICALL Java_Engine_debugGraph(JNIEnv *env, jclass cls, jfloat value, jint color)
	{
	gi.DebugGraph(value, color);
	}


// ------ helper function not based on Quake II Game interface

static jstring JNICALL Java_Engine_getGamePath(JNIEnv *env, jclass cls)
	{
	return (*env)->NewStringUTF(env, global_gameDirName);
	}

File: quake2/q2java/classes/q2java/Engine.java

package q2java;

public class Engine
	{   
	public final static int PRINT_LOW		= 0;	// pickup messages
	public final static int PRINT_MEDIUM		= 1;	// death messages
	public final static int PRINT_HIGH 		= 2;	// critical messages
	public final static int PRINT_CHAT 		= 3;

	public final static int MULTICAST_ALL 	= 0;
	public final static int MULTICAST_PHS 	= 1;
	public final static int MULTICAST_PVS 	= 2;
	public final static int MULTICAST_ALL_R 	= 3;
	public final static int MULTICAST_PHS_R 	= 4;
	public final static int MULTICAST_PVS_R 	= 5;

	//
	// per-level limits
	//
	public final static int MAX_CLIENTS     = 256;  // absolute limit
	public final static int MAX_EDICTS      = 1024; // must change protocol to increase more
	public final static int MAX_LIGHTSTYLES = 256;
	public final static int MAX_MODELS      = 256;  // these are sent over the net as bytes
	public final static int MAX_SOUNDS      = 256;  // so they cannot be blindly increased
	public final static int MAX_IMAGES      = 256;
	public final static int MAX_ITEMS       = 256;

	//
	// config strings are a general means of communication from
	// the server to all connected clients.
	// Each config string can be at most MAX_QPATH characters.
	//
	public final static int CS_NAME         	= 0;
	public final static int CS_CDTRACK  		= 1;
	public final static int CS_SKY      		= 2;
	public final static int CS_SKYAXIS  		= 3;        // %f %f %f format
	public final static int CS_SKYROTATE    	= 4;
	public final static int CS_STATUSBAR    	= 5;        // display program string

	public final static int CS_MAPCHECKSUM 	= 31;    // for catching cheater maps

	public final static int CS_MODELS           = 32;
	public final static int CS_SOUNDS           = (CS_MODELS+MAX_MODELS);
	public final static int CS_IMAGES           = (CS_SOUNDS+MAX_SOUNDS);
	public final static int CS_LIGHTS           = (CS_IMAGES+MAX_IMAGES);
	public final static int CS_ITEMS            = (CS_LIGHTS+MAX_LIGHTSTYLES);
	public final static int CS_PLAYERSKINS      = (CS_ITEMS+MAX_ITEMS);
	public final static int MAX_CONFIGSTRINGS   = (CS_PLAYERSKINS+MAX_CLIENTS);

	
	
	// protocol bytes that can be directly added to messages
	public final static int SVC_MUZZLEFLASH		= 1;
	public final static int SVC_MUZZLEFLASH2		= 2;
	public final static int SVC_TEMP_ENTITY		= 3;
	public final static int SVC_LAYOUT			= 4;
	public final static int SVC_INVENTORY		= 5;
	
	//
	// muzzle flashes / player effects
	//
	public final static int MZ_BLASTER      = 0;
	public final static int MZ_MACHINEGUN       = 1;
	public final static int MZ_SHOTGUN      = 2;
	public final static int MZ_CHAINGUN1        = 3;
	public final static int MZ_CHAINGUN2        = 4;
	public final static int MZ_CHAINGUN3        = 5;
	public final static int MZ_RAILGUN      = 6;
	public final static int MZ_ROCKET           = 7;
	public final static int MZ_GRENADE      = 8;
	public final static int MZ_LOGIN            = 9;
	public final static int MZ_LOGOUT           = 10;
	public final static int MZ_RESPAWN      = 11;
	public final static int MZ_BFG          = 12;
	public final static int MZ_SSHOTGUN         = 13;
	public final static int MZ_HYPERBLASTER     = 14;
	public final static int MZ_ITEMRESPAWN  = 15;
	public final static int MZ_SILENCED         = 128;      // bit flag ORed with one of the above numbers      

	// boxEntity() can return a list of either solid or trigger entities
	public final static int AREA_SOLID			= 1;
	public final static int AREA_TRIGGERS		= 2;	
	
	private final static int CALL_MULTICAST 		= 0;
	private final static int CALL_WRITEPOSITION	= 1;
	private final static int CALL_WRITEDIR 		= 2;
	private final static int CALL_WRITECHAR 		= 3;
	private final static int CALL_WRITEBYTE 		= 4;
	private final static int CALL_WRITESHORT 	= 5;
	private final static int CALL_WRITELONG 		= 6;
	private final static int CALL_WRITEFLOAT 	= 7;
	private final static int CALL_WRITEANGLE 	= 8;
	private final static int CALL_WRITESTRING 	= 9;
	private final static int CALL_UNICAST 		= 10;
	private final static int CALL_INPVS 			= 11;
	private final static int CALL_INPHS 			= 12;
	

	public native static void addCommandString(String s);


	public native static boolean areasConnected(int area1, int area2);


	public native static int argc();


	public native static String args();


	public native static String argv(int n);

/**
 * This method was created by a SmartGuide.
 * @return NativeEntity[]
 * @param mins Vec3
 * @param maxs Vec3
 * @param maxCount int
 * @param areaType int
 */
public static NativeEntity[] boxEntities(Vec3 mins, Vec3 maxs, int areaType) 
	{
	return boxEntities0(mins.x, mins.y, mins.z, maxs.x, maxs.y, maxs.z, areaType);
	}

/**
 * This method was created by a SmartGuide.
 * @return NativeEntity[]
 * @param mins Vec3
 * @param maxs Vec3
 * @param maxCount int
 * @param areaType int
 */
private native static NativeEntity[] boxEntities0(float minsx, float minsy, float minsz,
	float maxsx, float maxsy, float maxsz, int areaType);


	public native static void bprint(int printLevel, String s);


	public native static void configString(int num, String s);


	public native static void debugGraph(float value, int color);


	//
	// Methods that natively call back to the Quake2 Program
	//

	public native static void dprint(String s);


	public native static void error(String s);


/**
 * This method was created by a SmartGuide.
 * @return java.lang.String
 */
public native static String getGamePath();


	public native static int imageIndex(String name);


	private native static boolean inP0(float x1, float y1, float z1, float x2, float y2, float z2, int calltype);

	public static boolean inPHS(Vec3 p1, Vec3 p2)
		{
		return inP0(p1.x, p1.y, p1.z, p2.z, p2.y, p2.z, CALL_INPHS);
		}
	public static boolean inPVS(Vec3 p1, Vec3 p2)
		{
		return inP0(p1.x, p1.y, p1.z, p2.z, p2.y, p2.z, CALL_INPVS);
		}

	public native static int modelIndex(String name);

	public static void multicast(Vec3 origin, int to)
		{
		write0(null, origin.x, origin.y, origin.z, to, CALL_MULTICAST);
		}
	public static int pointContents(Vec3 point)
		{
		return pointContents0(point.x, point.y, point.z);
		}

	private native static int pointContents0(float x, float y, float z);


	public native static void setAreaPortalState(int portalnum, boolean open);


	public native static int soundIndex(String name);

/**
 * This method was created by a SmartGuide.
 * @return TraceResults
 * @param start Vec3
 * @param mins Vec3
 * @param maxs Vec3
 * @param end Vec3
 * @param passEnt NativeEntity
 * @param contentMask int
 */
public static TraceResults trace(Vec3 start, Vec3 mins, Vec3 maxs, Vec3 end, NativeEntity passEnt, int contentMask) 
	{
	return trace0(start.x, start.y, start.z, mins.x, mins.y, mins.z, maxs.x, maxs.y, maxs.z, end.x, end.y, end.z, passEnt, contentMask);
	}

/**
 * This method was created by a SmartGuide.
 * @return TraceResults
 * @param start Vec3
 * @param mins Vec3
 * @param maxs Vec3
 * @param end Vec3
 * @param passEnt NativeEntity
 * @param contentMask int
 */
private native static TraceResults trace0(float startx, float starty, float startz,
	float minsx, float minsy, float minsz, 
	float maxsx, float maxsy, float maxsz,
	float endx, float endy, float endz,
	NativeEntity passEnt, int contentMask);

public static void unicast(NativeEntity ent, boolean reliable)
	{
	write0(ent, 0, 0, 0, (reliable ? 1 : 0), CALL_UNICAST);
	}

	private native static void write0(Object obj, float x, float y, float z, int c, int calltype);

	public static void writeAngle(float f)
		{
		write0(null, f, 0, 0, 0, CALL_WRITEANGLE);
		}
	public static void writeByte(int c)
		{
		write0(null, 0, 0, 0, c, CALL_WRITEBYTE);
		}
	// wasn't used in original game dll, 
	// not sure if this should be a java char instead of int
	public static void writeChar(int c)         
		{
		write0(null, 0, 0, 0, c, CALL_WRITECHAR);
		}
	public static void writeDir(Vec3 pos)
		{
		write0(null, pos.x, pos.y, pos.z, 0, CALL_WRITEDIR);
		}
	public static void writeFloat(float f)
		{
		write0(null, f, 0, 0, 0, CALL_WRITEFLOAT);
		}
	public static void writeLong(int c)
		{
		write0(null, 0, 0, 0, c, CALL_WRITELONG);
		}
	public static void writePosition(Vec3 pos)
		{
		write0(null, pos.x, pos.y, pos.z, 0, CALL_WRITEPOSITION);
		}
	public static void writeShort(int c)
		{
		write0(null, 0, 0, 0, c, CALL_WRITESHORT);
		}
	public static void writeString(String s)
		{
		write0(s, 0, 0, 0, 0, CALL_WRITESTRING);
		}
}